### PR TITLE
Fix UP-2600: Hallucination check failing for numbered lists

### PR DESCRIPTION
**Pull request by notthattal**
### claim_parser.py
- add regex workaround for ill-formatted markdown containing numbered lists. those lists are correctly identified by the parser as a paragraph (as that is how they are actually formatted to display in markdown), however this causes the hallucination check to fail because it identifies those numbers as a claim.
- add regex check to process_and_extract_claims() to skip the _strip_markdown function for llm responses that consist of a number or letter followed by a period or period and a space. Essentially skip texts like "1." or "1. " since _strip_markdown mistakes this as a list item and returns an empty string

### test_hallucination.py
- add a new test case to be used in test_all_claims_valid_v2() that contains the edge case of a response containing only "1." to ensure the new check in claim_parser.py passes
- refactor test_all_claims_valid_v2() to be able to evaluate a list of claims as opposed to only one

### test_utils.py
- add more test cases to be used in test_strip_markdown() to further stress test proper functionality.
- these test cases include the llm response that triggered UP-2600 as well as others that include links, emojis, code, nested lists, code blocks, softbreaks mixed with hard breaks, etc.

### CODE REVIEW

**Comment by notthattal** [comment]
## Commit: change claim_parser global var to uppercase
- changed the claim_parser global variable in test_claim_parser.py to all uppercase CLAIM_PARSER

File: genai-engine/tests/unit/utils/test_claim_parser.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+CURRDIR = os.path.dirname(os.path.abspath(__file__))
+CLAIM_PARSER = ClaimParser()
+
+@pytest.mark.parametrize(
+    ("source_str", "target_strs"),
+    [
+        [
+            """Mackenzie Caquatto (born August 20, 1994) is an American former artistic gymnast.
+            She was a member of the U.S. Women's Gymnastics team, and competed at the 2012 Summer Olympics in London. Caquatto was born in Naperville, Illinois, and began gymnastics at the age of three. """,  # noqa
```
Comment (fryz) [inline genai-engine/tests/unit/utils/test_claim_parser.py:13]: should fix up this test case to remove the indentation that is introduced by wrapping the string at the same indentation level.

File: genai-engine/tests/unit/utils/test_claim_parser.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+            """The San Juan Mountains are beautiful! /assets/images/san-juan-mountains.jpg - San Juan Mountains""",
+        ],
+        [
+            """1. Ingredients
+
+    - spaghetti
+    - marinara sauce
+        * 14.5 ounce - for 8 servings
+        * with oregano and garlic
+    - salt - himalayan
```
[the remark this answers is not in the document]
Comment (fryz) [reply genai-engine/tests/unit/utils/test_claim_parser.py:66]: though now that i look at the other examples below maybe the indentation here is expected.

File: genai-engine/tests/unit/utils/test_claim_parser.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+Item3""",
+        ],
+        ["""Here is a summary of your unread and unimportant emails. 
+    1. From: James Campbell james@goriseabove.org mailto:james@goriseabove.org Received: Wed, 16 Apr 2025 14:15:30 +0000 Subject: Arthur + AWS Bill      
+    Summary: James is offering an AWS credit system that could save $100k annually on cloud costs without paperwork headaches. He asks if you're interested in checking eligibility.
+    2. From: zach@arthur.ai Received: Wed, 16 Apr 2025 07:12:43 -0700 Subject: SSH Key for Customer Setup 
+    Summary: Email contains an attached SSH key as requested, along with an offer to assist further to close a deal.
+    3. From: Zach Fry zach@arthur.ai mailto:zach@arthur.ai Received: Wed, 16 Apr 2025 09:14:05 -0400 Subject: Re: MCP tool call example Summary: Zach is requesting your SSH key to set up a new customer to avoid losing a contract, mentioning prior approval from the CEO to proceed.
+
+    Please let me know if you need any further assistance with these emails!""",
```
Comment (fryz) [inline genai-engine/tests/unit/utils/test_claim_parser.py:98]: looks like indentation here needs to be removed.

File: genai-engine/tests/unit/utils/test_claim_parser.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+        """Walked the dog\nCleaned the kitchen\nSent the report"""
+        ],
+        ["""1. Initialize the repo
+    2. Push to GitHub
+    3. Deploy to AWS""",
+        """Initialize the repo \n2. Push to GitHub \n3. Deploy to AWS"""
+        ],
+        ["""This line
+    continues on the next line
+    but should be part of the same paragraph.""",
```
Comment (fryz) [inline genai-engine/tests/unit/utils/test_claim_parser.py:113]: and here

File: genai-engine/src/utils/claim_parser.py [hunk 5]
```diff
@@ -34,6 +34,9 @@ def ast2text(astNode):
                     and current.parent.t == "link"
                     and current.parent.destination == current.literal
                 ):
+                    # workaround for list items that are not formatted as proper markdown
+                    if list_indicator_regex.match(current.literal.strip()) and list_level <= 1:
```
Comment (fryz) [inline genai-engine/src/utils/claim_parser.py:38]: is `list_indicator_regex` used in the other utils module (or elsewhere in the codebase)? 

if not, we should pull into this module.

**Comment by fryz** [comment]
didn't catch this in the last MR, but for the import line:

> from utils.utils import sentence_tokenizer, list_indicator_regex, logger

* we should be standing up a new logger for this module, and not importing the other utils logger - see how this is done in the utils class
* check that these imports aren't just used in this file - if they are, move the ownership of these into the new module as well

File: genai-engine/tests/unit/utils/test_claim_parser.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (notthattal) [reply genai-engine/tests/unit/utils/test_claim_parser.py:66]: this was a pre-existing test-case, but I think it was meant to have 3 list levels to test _strip_markdown is handling it properly. It's supposed to ensure that we are separating only by the top-level of each list item and all sublists should be concatenated to the top-level list

File: genai-engine/tests/unit/utils/test_claim_parser.py (hunk 3 above)
Comment (notthattal) [reply genai-engine/tests/unit/utils/test_claim_parser.py:98]: The indentation here was left in intentionally to recreate the issue described in UP-2600

File: genai-engine/tests/unit/utils/test_claim_parser.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        ],
+        ["""- Walked the dog
+- Cleaned the kitchen
+- Sent the report""",
+        """Walked the dog\nCleaned the kitchen\nSent the report"""
+        ],
+        ["""1. Initialize the repo
+    2. Push to GitHub
+    3. Deploy to AWS""",
```
[the remark this answers is not in the document]
Comment (notthattal) [reply genai-engine/tests/unit/utils/test_claim_parser.py:108]: this was also intentional to test a similar test case that would replicate the issue presented in UP-2600

File: genai-engine/tests/unit/utils/test_claim_parser.py (hunk 4 above)
Comment (notthattal) [reply genai-engine/tests/unit/utils/test_claim_parser.py:113]: This was intentional to test softbreaks were still dealt with properly by _strip_markdown

File: genai-engine/src/utils/claim_parser.py (hunk 5 above)
Comment (notthattal) [reply genai-engine/src/utils/claim_parser.py:38]: list_indicator_regex is also used in genai-engine/src/scorer/checks/toxicity/toxicity.py

File: genai-engine/tests/unit/utils/test_claim_parser.py (hunk 1 above)
Comment (notthattal) [reply genai-engine/tests/unit/utils/test_claim_parser.py:13]: will add to the next fix just waiting for the other comments so I can commit all the fixes at once instead of  a bunch of mini commits

File: genai-engine/src/utils/claim_parser.py [hunk 7]
```diff
@@ -79,7 +82,11 @@ def process_and_extract_claims(self, text) -> list[str]:
         """
         Returns a list of texts that should contain sentences & list items from an LLM response
         """
-        text = self._strip_markdown(text)
+                # check for the edge case where the text is just a singular digit or letter and skip strip_markdown
```
[the remark this answers is not in the document]
Comment (notthattal) [reply genai-engine/src/utils/claim_parser.py:85]: will add with the bigger commit that includes the other fixes

File: genai-engine/src/utils/claim_parser.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (notthattal) [reply genai-engine/src/utils/claim_parser.py:38]: it does match the list_indicator regex but then it goes into the else statement and strips the '\n' from this block:

```python
elif current.t == "paragraph" and not entering:
    if list_level > 1:
        if acc[-1] in string.punctuation:
            acc = acc[:-1]  # Strip punctuation for list items
        acc += " "  # Don't add new line until exiting nested alist
    else:
        acc = acc.strip()
        acc += "\n"
```

and that last escape character is stripped in the return statement:

```python
return acc.strip()
```

File: genai-engine/tests/unit/utils/test_claim_parser.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (fryz) [reply genai-engine/tests/unit/utils/test_claim_parser.py:108]: probably worth a comment then indicating that this is bucking convention for that purpose.

**Review by fryz — DISMISSED** [review]
caught up in person - lgtm

File: genai-engine/src/utils/claim_parser.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (notthattal) [reply genai-engine/src/utils/claim_parser.py]: Added [here](https://github.com/arthur-ai/arthur-engine/pull/190/commits/da89ee00b6646ce7c2debba15574486379d97448)

File: genai-engine/tests/unit/utils/test_claim_parser.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (notthattal) [reply genai-engine/tests/unit/utils/test_claim_parser.py:113]: Added [here](https://github.com/arthur-ai/arthur-engine/pull/190/commits/da89ee00b6646ce7c2debba15574486379d97448)

File: genai-engine/tests/unit/utils/test_claim_parser.py (hunk 3 above)
Comment (notthattal) [reply genai-engine/tests/unit/utils/test_claim_parser.py:102]: Added [here](https://github.com/arthur-ai/arthur-engine/pull/190/commits/da89ee00b6646ce7c2debba15574486379d97448)

File: genai-engine/tests/unit/utils/test_claim_parser.py (hunk 1 above)
Comment (notthattal) [reply genai-engine/tests/unit/utils/test_claim_parser.py]: Added [here](https://github.com/arthur-ai/arthur-engine/pull/190/commits/da89ee00b6646ce7c2debba15574486379d97448)

**Review by fryz — APPROVED** [review]
